### PR TITLE
[TTS]Cantonese FastSpeech2 Training, test=tts

### PR DESCRIPTION
**Pull request by WongLaw**
Cantonese FastSpeech2 Training, test=tts

### CODE REVIEW

File: examples/canton/tts3/local/preprocess.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+stage=1
+stop_stage=100
+
+config_path=$1
+
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+    # get durations from MFA's result
+    echo "Generate durations.txt from MFA results ..."
+    python3 ${MAIN_ROOT}/utils/gen_duration_from_textgrid.py \
+        --inputdir=./canton_alignment \
```
Comment (yt605155624) [inline examples/canton/tts3/local/preprocess.sh:12]: 这个 mfa 的结果是否可以像 csmsc/tts3 一样打包一份把链接放到 readme?

File: examples/canton/tts3/README.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    ├── raw
+    └── speech_stats.npy
+```
+The dataset is split into 3 parts, namely `train`, `dev`, and` test`, each of which contains a `norm` and `raw` subfolder. The raw folder contains speech、pitch and energy features of each utterance, while the norm folder contains normalized ones. The statistics used to normalize features are computed from the training set, which is located in `dump/train/*_stats.npy`.
+
+Also, there is a `metadata.jsonl` in each subfolder. It is a table-like file that contains phones, text_lengths, speech_lengths, durations, the path of speech features, the path of pitch features, a path of energy features, speaker, and id of each utterance.
+
+### Training details can refer to the script of examples/aishell3/tts3.
+
+## Pretrained Model(Waiting========)
```
Comment (yt605155624) [inline examples/canton/tts3/README.md:77]: 现在模型已经训练完了吧？感觉预训练模型可以放上来了

File: examples/canton/tts3/README.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+├── default.yaml            # default config used to train fastspeech2
+├── energy_stats.npy        # statistics used to normalize energy when training fastspeech2
+├── phone_id_map.txt        # phone vocabulary file when training fastspeech2
+├── pitch_stats.npy         # statistics used to normalize pitch when training fastspeech2
+├── snapshot_iter_96400.pdz # model parameters and optimizer states
+├── speaker_id_map.txt      # speaker id map file when training a multi-speaker fastspeech2
+└── speech_stats.npy        # statistics used to normalize spectrogram when training fastspeech2
+```
+You can use the following scripts to synthesize for `${BIN_DIR}/../sentences.txt` using pretrained fastspeech2 and parallel wavegan models.
+```bash
```
Comment (yt605155624) [inline examples/canton/tts3/README.md:96]: 这里等待前端写好之后再更新

File: examples/canton/tts3/path.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (yt605155624) [inline examples/canton/tts3/path.sh]: 这个文件可以软链接 csmsc/tts3/path.sh

File: paddlespeech/t2s/exps/fastspeech2/preprocess.py [hunk 5]
```diff
@@ -54,8 +54,15 @@ def process_sentence(config: Dict[str, Any],
     record = None
     if utt_id in sentences:
         # reading, resampling may occur
-        wav, _ = librosa.load(str(fp), sr=config.fs)
-        if len(wav.shape) != 1:
+        wav, _ = librosa.load(
+            str(fp), sr=config.fs,
+            mono=False) if "canton" in str(fp) else librosa.load(
+                str(fp), sr=config.fs)
+        if len(wav.shape) == 2 and "canton" in str(fp):
```
Comment (yt605155624) [inline paddlespeech/t2s/exps/fastspeech2/preprocess.py:61]: mfa 的时候要求数据集直接放在 datasets/ 里面，但是此处又要求放在  datasets/canton_all 里面，此处统一下吧，如果这里不好改就改 mfa

File: examples/canton/tts3/README.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+To obtain better performance, please combine these two datasets together as follows:
+
+```bash
+mkdir -p ~/datasets/canton_all/WAV
+cp -r ~/datasets/Guangzhou_Cantonese_Scripted_Speech_Corpus_Daily_Use_Sentence/WAV/* ~/datasets/canton_all/WAV
+cp -r ~/datasets/Guangzhou_Cantonese_Scripted_Speech_Corpus_in_Vehicle/WAV/* ~/datasets/canton_all/WAV
+```
+
+After that, it should be look like:
+~/datasets/canton_all_
```
Comment (yt605155624) [inline examples/canton/tts3/README.md]: 这里格式错误，用 
`
```text 
```
`
包起来

File: examples/canton/tts3/README.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
+# FastSpeech2 with Cantonese language
+This example contains code used to train a [Fastspeech2](https://arxiv.org/abs/2006.04558) model with [Guangzhou_Cantonese_Scripted_Speech_Corpus_Daily_Use_Sentence](https://magichub.com/datasets/guangzhou-cantonese-scripted-speech-corpus-daily-use-sentence/) and [Guangzhou_Cantonese_Scripted_Speech_Corpus_in_Vehicle](https://magichub.com/datasets/guangzhou-cantonese-scripted-speech-corpus-in-the-vehicle/).
+
+
+fastspeech2 model here.
```
Comment (yt605155624) [inline examples/canton/tts3/README.md]: del

File: examples/canton/tts3/README.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+## Dataset
+### Download and Extract
+If you don't have the Cantonese datasets mentioned above, please download [Guangzhou_Cantonese_Scripted_Speech_Corpus_Daily_Use_Sentence](https://magichub.com/datasets/guangzhou-cantonese-scripted-speech-corpus-daily-use-sentence/) and [Guangzhou_Cantonese_Scripted_Speech_Corpus_in_Vehicle](https://magichub.com/datasets/guangzhou-cantonese-scripted-speech-corpus-in-the-vehicle/) under `~/datasets/`.
```
Comment (yt605155624) [inline examples/canton/tts3/README.md]: please download and unzip，加上 unzip

File: examples/canton/tts3/README.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+│   └── WAV
+│       └──G0001
+│       └──G0002
+│       ...
+│       └──G0071
+│       └──G0072
+
+ 
+### Get MFA Result and Extract
+We use [MFA2.x](https://github.com/MontrealCorpusTools/Montreal-Forced-Aligner) to get durations for aishell3_fastspeech2.
```
Comment (yt605155624) [inline examples/canton/tts3/README.md]: 咱们这块用的是 MFA 1.X, aishell3_fastspeech2 也错了，可以改成 conton_fastspeech2

File: examples/canton/tts3/README.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+You can train your MFA model reference to [canton_mfa example](https://github.com/PaddlePaddle/PaddleSpeech/tree/develop/examples/other/mfa) (use MFA1.x now) of our repo.
+
```
Comment (yt605155624) [inline examples/canton/tts3/README.md:31]: 增加我们处理好的 canton_alignment.zip 的链接

File: examples/canton/tts3/README.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    ├── norm
+    ├── pitch_stats.npy
+    ├── raw
+    └── speech_stats.npy
+```
+The dataset is split into 3 parts, namely `train`, `dev`, and` test`, each of which contains a `norm` and `raw` subfolder. The raw folder contains speech、pitch and energy features of each utterance, while the norm folder contains normalized ones. The statistics used to normalize features are computed from the training set, which is located in `dump/train/*_stats.npy`.
+
+Also, there is a `metadata.jsonl` in each subfolder. It is a table-like file that contains phones, text_lengths, speech_lengths, durations, the path of speech features, the path of pitch features, a path of energy features, speaker, and id of each utterance.
+
+### Training details can refer to the script of examples/aishell3/tts3.
```
Comment (yt605155624) [inline examples/canton/tts3/README.md]: examples/aishell3/tts3 变成链接形式，如 `[examples/aishell3/tts3](../../aishell3/tts3)`

File: examples/canton/tts3/local/train.sh [hunk 12]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (yt605155624) [inline examples/canton/tts3/local/train.sh]: 直接软链接 aishell3/local/train.sh 就行吧

File: examples/canton/tts3/local/synthesize.sh [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        --am=fastspeech2_aishell3 \
+        --am_config=${config_path} \
+        --am_ckpt=${train_output_path}/checkpoints/${ckpt_name} \
+        --am_stat=dump/train/speech_stats.npy \
+        --voc=pwgan_aishell3 \
+        --voc_config=pwg_aishell3_ckpt_0.5/default.yaml \
+        --voc_ckpt=pwg_aishell3_ckpt_0.5/snapshot_iter_1000000.pdz \
+        --voc_stat=pwg_aishell3_ckpt_0.5/feats_stats.npy \
+        --test_metadata=dump/test/norm/metadata.jsonl \
+        --output_dir=${train_output_path}/test_new \
```
Comment (yt605155624) [inline examples/canton/tts3/local/synthesize.sh]: 这里 test_new 改成 test 后和 aishell3/local/synthesize.sh 一模一样，可以直接软链接过来

File: examples/canton/tts3/conf/default.yaml [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+window: "hann"     # Window function.
+
+# Only used for feats_type != raw
+
+fmin: 80           # Minimum frequency of Mel basis.
+fmax: 7600         # Maximum frequency of Mel basis.
+n_mels: 80         # The number of mel basis.
+
+# Only used for the model using pitch features (e.g. FastSpeech2)
+f0min: 110          # Minimum f0 for pitch extraction.
```
Comment (yt605155624) [inline examples/canton/tts3/conf/default.yaml:22]: 这里如果觉得需要的话，可以说明下为什么比其他数据集的 f0min 要高

目前 All frames seems to be unvoiced. 的报错还是有点多，这个数值再看看吧

File: examples/canton/tts3/local/preprocess.sh (hunk 1 above)
Comment (WongLaw) [reply examples/canton/tts3/local/preprocess.sh:12]: done

File: examples/canton/tts3/README.md (hunk 9 above)
Comment (WongLaw) [reply examples/canton/tts3/README.md]: done

File: examples/canton/tts3/README.md (hunk 10 above)
Comment (WongLaw) [reply examples/canton/tts3/README.md:31]: done

File: examples/canton/tts3/README.md (hunk 8 above)
Comment (WongLaw) [reply examples/canton/tts3/README.md]: done

File: examples/canton/tts3/README.md (hunk 7 above)
Comment (WongLaw) [reply examples/canton/tts3/README.md]: done

File: examples/canton/tts3/README.md (hunk 6 above)
Comment (WongLaw) [reply examples/canton/tts3/README.md]: done

File: examples/canton/tts3/README.md (hunk 10 above)
Comment (yt605155624) [reply examples/canton/tts3/README.md:31]: 链接的名称改成 canton_alignment.zip 吧

File: examples/canton/tts3/local/preprocess.sh [hunk 15]
```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+
+stage=1
```
Comment (yt605155624) [inline examples/canton/tts3/local/preprocess.sh]: stage=0, 否则会跳过生成 duration.txt 的步骤

File: examples/canton/tts3/README.md (hunk 10 above)
Comment (WongLaw) [reply examples/canton/tts3/README.md:31]: done

**Comment by yt605155624** [comment]
- https://github.com/PaddlePaddle/PaddleSpeech/blob/66a9cf8ebc0be8deb68c6fa9b53f961aa647b045/paddlespeech/t2s/exps/fastspeech2/preprocess.py#L104 对于 f0 全为 0 的情况 return None 进行过滤
- https://github.com/PaddlePaddle/PaddleSpeech/blob/66a9cf8ebc0be8deb68c6fa9b53f961aa647b045/paddlespeech/t2s/datasets/get_feats.py#L105 这里 print 信息改成 All frames seems to be unvoiced, this utt will be removed.

File: examples/canton/tts3/conf/default.yaml [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+fs: 24000          # sr
+n_fft: 2048        # FFT size (samples).
+n_shift: 300       # Hop size (samples). 12.5ms
+win_length: 1200   # Window length (samples). 50ms
+                   # If set to null, it will be the same as fft_size.
+window: "hann"     # Window function.
+
+# Only used for feats_type != raw
+
+fmin: 80           # Minimum frequency of Mel basis.
```
Comment (yt605155624) [inline examples/canton/tts3/conf/default.yaml:14]: fmin 是不是也要改成 110，不然感觉会包含了噪声

File: examples/canton/tts3/path.sh (hunk 4 above)
Comment (WongLaw) [reply examples/canton/tts3/path.sh]: done

File: examples/canton/tts3/README.md (hunk 11 above)
Comment (WongLaw) [reply examples/canton/tts3/README.md]: done

File: examples/canton/tts3/local/train.sh (hunk 12 above)
Comment (WongLaw) [reply examples/canton/tts3/local/train.sh]: done

File: examples/canton/tts3/local/synthesize.sh (hunk 13 above)
Comment (WongLaw) [reply examples/canton/tts3/local/synthesize.sh]: done

File: examples/canton/tts3/conf/default.yaml (hunk 14 above)
Comment (WongLaw) [reply examples/canton/tts3/conf/default.yaml:22]: done

File: examples/canton/tts3/local/preprocess.sh (hunk 15 above)
Comment (WongLaw) [reply examples/canton/tts3/local/preprocess.sh]: done

**Comment by yt605155624** [comment]
> * https://github.com/PaddlePaddle/PaddleSpeech/blob/66a9cf8ebc0be8deb68c6fa9b53f961aa647b045/paddlespeech/t2s/exps/fastspeech2/preprocess.py#L104
>    对于 f0 全为 0 的情况 return None 进行过滤
> * https://github.com/PaddlePaddle/PaddleSpeech/blob/66a9cf8ebc0be8deb68c6fa9b53f961aa647b045/paddlespeech/t2s/datasets/get_feats.py#L105
>    这里 print 信息改成 All frames seems to be unvoiced, this utt will be removed.

resolved

File: examples/canton/tts3/conf/default.yaml (hunk 16 above)
Comment (yt605155624) [reply examples/canton/tts3/conf/default.yaml:14]: 就现在合成效果来看这块问题不大，改的话和预训练的 voc 参数不匹配，可能影响合成效果，这个之后再看

File: examples/canton/tts3/run.sh [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    # synthesize, vocoder is pwgan by default
+    CUDA_VISIBLE_DEVICES=${gpus} ./local/synthesize.sh ${conf_path} ${train_output_path} ${ckpt_name} || exit -1
+fi
+
+if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
+    # synthesize_e2e, vocoder is pwgan by default
+    CUDA_VISIBLE_DEVICES=${gpus} ./local/synthesize_e2e.sh ${conf_path} ${train_output_path} ${ckpt_name} || exit -1
+fi
+
+if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ]; then
```
Comment (yt605155624) [inline examples/canton/tts3/run.sh:39]: 加入文本前端之后后面这些 stage 可以删掉

**Review by yt605155624 — APPROVED** [review]
LGTM
